### DISCUSSION TITLE
Things that are coming some time #77

### DISCUSSION THREAD

**Discussion by Miraculix200** [SpenceKonde/DxCore#78] [Future development discussion]
Re: https://github.com/SpenceKonde/DxCore/issues/77

Nice. Looking forward to the sleepLib. I guess I'll find some use for it in the future.

Would be nice to also have millis()/micros()/delay() work properly while using CLKCTRL to reduce CPU speed, so external libraries (or Serial library) don't produce a mess when using it. Not sure if it's even possible? I gave up after some hours.

**Comment by SpenceKonde** [comment]
That'd be pretty cool wouldn't it? but seeing as everything else also breaks the moment F_CPU is changed at runtime (the code is full of references to the F_CPU define, and rewriting them to be non-constant would force it to keep more branches in to handle different speeds. 

megaTinyCore only supports operating on a single, fixed system clock speed. Supporting non-constant F_CPU would not only take a huge amount of development effort, but necessarily have an adverse impact on execution speed and binary size. And third party libraries still wouldn't work because THEY are all written for constant F_CPU. Besides, for power consumption, you're usually better running at a normal speed for short bursts and sleeping vs running at a low clock continuously, so it's not clear why this would even be desirable.. 

exposing a function to switch millis to the RTC and back (which would be part of sleepLib anyway) would give you the ability to switch those timing functions over if you insisted on doing this, but I am not planning to go down the path of trying to give proper support for this

**Reply by Miraculix200** [reply]
I thought so, that F_CPU was the main problem. So it's not worth the effort. 

Anyway, I thought as you have a better overview of the core than me, you may have an idea how to make it work. 

Why it's desirable? Well because it's theoretically possible haha. AVR MCUs are often used because they are low power devices (it's the main reason I often use it instead of a Espressif MCU). So saving a few mA here and there doesn't hurt. But sleeping the device is much more important. 

Basically I just wanted to change the CPU speed in code so I don't have to change the CPU speed option in the IDE all the time. Right now I simply add a #error directive to remind me to change the CPU speed to the lowest speed necessary for the code. And I don't really have the need to increase the CPU speed during runtime.

**Comment by SpenceKonde** [comment]
To deal with the loss of serial, though? To support tuning (here to get 24 MHz internal and like, whatever other weirdo speeds you want), but more importantly classic ATTinyCore), I'm going to need a way to get data out for debugging as well as general data gathering. 

Am planning to use another board as SPI slave to the one under test, and have it just blast out what it is sent over spi via serial. key is to make sure you don't miss SPI data, which is gonna make for an.... interesting piece of code.
